### ISSUE TITLE
Impute for Test set

### ISSUE THREAD

**Issue by prabhakaran2812** [pycaret/pycaret#2200]
**Is your feature request related to a problem? Please describe.**
Yes, I can impute data for train but was unable to do for test

**Describe the solution you'd like**
Setup enabling imputation for test separately. Alternate option is I could merge train & test and can impute. Which means I used setup just for imputation sake alone and then separated the test and train to run the setup against my train. I was essentially running setup twice to accomplish test impute

**Describe alternatives you've considered**
Explained above, ran setup a separate task for imputation

**Additional context**
I can more details, if any questions are specific. Not sure may be im already not aware of an existing feature. Please advise.

**Comment by ngupta23** [comment]
Why do you want to impute the test (unseen) data? The imputation should not use any information from test. This will cause leakage in your model and it will produce biased results.

The way it works is that

1. The imputation is done on the train dataset only. The imputation parameters are stored in the model that is built on this train data.
2. Then when you try to predict on the test data (which has missing values), these missing values will get imputed using the stored imputation parameters from the train step. So you do not need to specifically impute the test data yourself.

I hope that answers your question. If not, could you describe your use case in more detail and provide the code for what you are trying to do?

**Comment by prabhakaran2812** [comment]
Thanks for the reply. By imputation, do you mean prediction? Below is my scenario
I've x1,x2,x3,x4,y1 dataset
I've missing values in x in both train and test data set. I am using below setup
iterative_imputation_iters=10,numeric_iterative_imputer='lightgbm',categorical_iterative_imputer='lightgbm'
for my train dataset imputation as preprocess step.

Can I do similar preprocess imputation on test dataset? It will be very helpful if we have this option or am I missing something.

**Comment by ngupta23** [comment]
> By imputation, do you mean prediction?

No, I mean imputing the x* in your test set

> Can I do similar preprocess imputation on test dataset?

As I mentioned above, the value to impute for the missing x* rows will be determined using the train dataset. Once these are determined, they will be used to impute the missing x* values in the test set before the predictions can be done. So you do not have to do anything separately for the test set.

@tvdboom @Yard1 - Do you have any further comments on this?

**Comment by tvdboom** [comment]
@prabhakaran2812 What you are asking for already happens. The IterativeImputer class used to impute the training set is also used to impute the test set. What you are missing here is this. You said `I am using below setup ... for my train dataset imputation as preprocess step`. But it's not only the training set that gets transformed by preprocessing. The test set as well. The only difference is that the preprocessing pipeline (containing the Imputer in your case) is only **fitted on the training set**. After fitting, it **transforms the complete dataset** (train and test). So your test set is, as you asked, being imputed the same way as the training set.

You can check that there are indeed no missing values in the test set after setup, running `get_config("X_test")`

**Comment by prabhakaran2812** [comment]
Thanks @ngupta23 , @tvdboom 
The challenge is, when I use test_data=testdataset in setup, it is expecting the target column to be present in my test data. I don't have any target values in my test dataset. I can use dummy 0 values in my target data if thats what is the normal way to do. I wasn't sure. Please clarify.
